### PR TITLE
Fixed grid

### PR DESCRIPTION
**Pull request by arashr88**
## Description
Please include a summary of the change and which issue is fixed. Please 
also include relevant motivation and context. For error/bug reporting, 
please include the error and it's accompanying context.

Fixes # (issue)

## Type of Change
Please delete options that are not relevant. Please specify if change 
applies to GUI.

- [ * ] Bug fix (non-breaking change which fixes an issue)
- [ * ] New feature (non-breaking change which adds functionality)
- [ * ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ * ] This change requires a documentation update

## How has this change been tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Also, list any relevant details for your test configuration.

Update the tests for fixed-grid resource allocation

**Test Configuration**:
* Operating System:
* Python version:

## Does change comply with project standards and guideline?
Standards and guidelines can be found on the [Project Homepage](https://github.com/SDNNetSim/SDON_simulator).
- [ ] Commit formatting
- [ ] Versioning
- [ ] Code style
- [ ] Coding guidelines
- [ ] Unit Testing

### CODE REVIEW

File: bash_scripts/run_sim.sh [hunk 1]
```diff
@@ -1,40 +1,68 @@
...
 #SBATCH -G 0
-#SBATCH --mem=16000
-#SBATCH -t 1-00:00:00
-#SBATCH -o slurm-%j.out
+#SBATCH --mem=40000
+#SBATCH -t 14-00:00:00
+#SBATCH -o slurm-%A_%a.out  # Output file for each task
+#SBATCH --array=0-3  # Define array size based on total combinations
 
-# This script is designed to run a non-artificial intelligence simulation on the Unity cluster at UMass Amherst.
```
Comment (ryanmccann1024) [inline bash_scripts/run_sim.sh:10]: I think we want to keep this comment to describe what this bash script is designed to do.

File: bash_scripts/run_sim.sh [hunk 2]
```diff
@@ -1,40 +1,68 @@
...
-# Users can provide custom parameters via the command line using -- before the parameter list.
+set -e  # Stop the script if any command fails
 
-# Ensure the script stops if any command fails
-set -e
+echo "Starting job with SLURM_ARRAY_TASK_ID: $SLURM_ARRAY_TASK_ID"
 
-# Change to the home directory and then to the SDN simulator directory
```
Comment (ryanmccann1024) [inline bash_scripts/run_sim.sh:16]: We should take this as command line input but have a default to this directory. It's not clear where the simulation would run for the user.

File: bash_scripts/run_sim.sh [hunk 3]
```diff
@@ -1,40 +1,68 @@
...
+# Load the Python 3.11.7 module
+module load python/3.11.7
 
-# Load the required Python module
-module load python/3.11.0
+# Verify the loaded Python version
+python --version
 
-# Activate the virtual environment
-source venvs/unity_venv/venv/bin/activate
```
Comment (ryanmccann1024) [inline bash_scripts/run_sim.sh:26]: I think the venv is only used for RL, just leaving a note so I can remember to double check.

File: bash_scripts/run_sim.sh [hunk 4]
```diff
@@ -1,40 +1,68 @@
...
+# Define parameter arrays
+allocation_methods=("first_fit" "last_fit")
+spectrum_allocation_priorities=("CSB" "BSC")
+cores_per_link=(4)
 
-# Install the required Python packages
-pip install -r requirements.txt
```
Comment (ryanmccann1024) [inline bash_scripts/run_sim.sh:29]: I think we should still do this as if we add another package and Unity does not have it the simulation will fail.

File: bash_scripts/run_sim.sh [hunk 5]
```diff
@@ -1,40 +1,68 @@
...
+# Define additional variables
+num_requests=15000
+multi_fiber=False
 
-# Register custom reinforcement learning environments in Stable Baselines 3
```
Comment (ryanmccann1024) [inline bash_scripts/run_sim.sh:31]: Now that I see this, it seems like this is a working version of an old `run_sim.sh` as this is not on main?

File: bash_scripts/run_sim.sh [hunk 6]
```diff
@@ -1,40 +1,68 @@
...
+# Print parameters for this task
+echo "Running simulation with:"
+echo "  Allocation method: $allocation_method"
+echo "  Spectrum priority: $spectrum_priority"
+echo "  Cores per link: $cores"
+echo "  Number of requests: $num_requests"
+echo "  Multi-fiber: $multi_fiber"
+
+# Move to the project directory
+cd ~/Git/SDON_simulator/ || { echo "Failed to change to SDON_simulator directory"; exit 1; }
```
Comment (ryanmccann1024) [inline bash_scripts/run_sim.sh:59]: My comment is related above we should have this as input but also have a default option.

File: data/raw/USB6014.txt [hunk 7]
```diff
@@ -0,0 +1,158 @@
+1	2	277.1
```
Comment (ryanmccann1024) [inline data/raw/USB6014.txt:1]: We should ask Scott to have some sort of doc on this, the more we add networks beyond the basics it'll be hard to distinguish them in 1-2 years from now.

File: data/raw/USB6014_core_nodes.txt [hunk 8]
```diff
@@ -0,0 +1,14 @@
+51
```
Comment (ryanmccann1024) [inline data/raw/USB6014_core_nodes.txt:1]: Same with this script, maybe we have a README in the raw data directory for now.

File: data_scripts/structure_data.py [hunk 9]
```diff
@@ -15,7 +15,7 @@ def assign_link_lengths(network_fp: str, node_pairs_dict: dict, constant_weight:
     with open(network_fp, 'r', encoding='utf-8') as file_obj:
         for line in file_obj:
             src, dest, link_len_str = line.strip().split('\t')
-            link_len = int(link_len_str) if not constant_weight else 1
+            link_len = float(link_len_str) if not constant_weight else 1
```
Comment (ryanmccann1024) [inline data_scripts/structure_data.py:18]: If we want this to be a float the `else` should be `1.0`

File: helper_scripts/spectrum_helpers.py [hunk 10]
```diff
@@ -22,9 +22,23 @@ def __init__(self, engine_props: dict, sdn_props: object, spectrum_props: object
 
     def _check_free_spectrum(self, link_tuple: tuple, rev_link_tuple: tuple):
         core_arr = self.sdn_props.net_spec_dict[link_tuple]['cores_matrix'][self.curr_band][self.core_num]
-        spectrum_set = core_arr[self.start_index:self.end_index + self.engine_props['guard_slots']]
         rev_core_arr = self.sdn_props.net_spec_dict[rev_link_tuple]['cores_matrix'][self.curr_band][self.core_num]
-        rev_spectrum_set = rev_core_arr[self.start_index:self.end_index + self.engine_props['guard_slots']]
+        if self.spectrum_props.slots_needed == 1 and self.engine_props['guard_slots'] == 0:
```
Comment (ryanmccann1024) [inline helper_scripts/spectrum_helpers.py:26]: I believe we wrote a test for this, if not, we definitely should.

File: helper_scripts/stats_helpers.py [hunk 11]
```diff
@@ -178,24 +195,40 @@ def get_blocking(self):
...
         self.stats_props.sim_block_list.append(blocking_prob)
-
+        self.stats_props.sim_bit_rate_block_list.append(bit_rate_blocking_prob)
     def _handle_iter_lists(self, sdn_data: object):
         for stat_key in sdn_data.stat_key_list:
             # TODO: Eventually change this name (sdn_data)
             curr_sdn_data = sdn_data.get_data(key=stat_key)
-
+            if stat_key == 'xt_list':
+                self.stats_props.xt_list.append(mean(curr_sdn_data)) # TODO: double-check
```
Comment (ryanmccann1024) [inline helper_scripts/stats_helpers.py:210]: I see TODO double check here, is it functioning fine?

File: helper_scripts/stats_helpers.py [hunk 12]
```diff
@@ -178,24 +195,40 @@ def get_blocking(self):
...
                 elif stat_key == 'modulation_list':
                     bandwidth = sdn_data.bandwidth_list[i]
+                    band = sdn_data.band_list[i]
                     self.stats_props.mods_used_dict[bandwidth][data] += 1
-                elif stat_key == 'xt_list':
-                    self.stats_props.xt_list.append(data)
+                    self.stats_props.mods_used_dict[data][band] += 1
+                    self.stats_props.mods_used_dict[data]['length'][band].append(sdn_data.path_weight)
+                    self.stats_props.mods_used_dict[data]['length']['overall'].append(sdn_data.path_weight)
+                # elif stat_key == 'xt_list':
```
Comment (ryanmccann1024) [inline helper_scripts/stats_helpers.py:221]: Comment should be deleted or not commented.

File: helper_scripts/stats_helpers.py [hunk 13]
```diff
@@ -178,24 +195,40 @@ def get_blocking(self):
...
+                #     self.stats_props.xt_list.append(data) # TODO: double-check
+                elif stat_key == 'start_slot_list':
```
Comment (ryanmccann1024) [inline helper_scripts/stats_helpers.py:223]: I think @kbempah  did something where we could loop over a list of items and access the member in a class with a string. It's probably worth looking into as this scales.

File: src/routing.py [hunk 14]
```diff
@@ -179,6 +183,50 @@ def _init_route_info(self):
...
+        
+        path_cnt = 0
+        for item in loaded_data_dict:
+            cnt = 0
+            first_element = item[5][0][0][0][0]
+            last_element = item[5][0][0][0][-1]
+            if first_element in src_des_list and last_element in src_des_list:
+                self.route_props.connection_index.append(path_cnt)
+                for path in item[5][0]:
+                    if cnt == 3:
```
Comment (ryanmccann1024) [inline src/routing.py:204]: Should comment why, it's not clear why 3 is important in this function.

File: src/routing.py [hunk 15]
```diff
@@ -179,6 +183,50 @@ def _init_route_info(self):
...
+        """
+        Load the k-shortest paths from an external file.
+        """
+        loaded_data_dict = np.load('USB6014-10SP.npy', allow_pickle=True)
+        src_des_list = [int(self.sdn_props.source), int(self.sdn_props.destination)]
+        
+        path_cnt = 0
+        for item in loaded_data_dict:
+            cnt = 0
+            first_element = item[5][0][0][0][0]
```
Comment (ryanmccann1024) [inline src/routing.py:199]: Is first element equal to the first node or first path?

File: src/routing.py [hunk 16]
```diff
@@ -179,6 +183,50 @@ def _init_route_info(self):
...
+            last_element = item[5][0][0][0][-1]
```
Comment (ryanmccann1024) [inline src/routing.py:200]: I think this should be commented as well, the indexing here is hard to follow. I think it's because `item` is a vague variable name and doesn't follow our naming guidelines as it looks like a matrix but I'm not sure.

File: src/routing.py [hunk 17]
```diff
@@ -179,6 +183,50 @@ def _init_route_info(self):
...
+                    if cnt == 3:
+                        break
+                    if first_element == int(self.sdn_props.source):
+                        temp_path = list(path[0])
+                    else:
+                        temp_path = list(path[0][::-1])
+
+                    
+                    temp_path = list(map(str, temp_path))
+                    path_len1 = find_path_len(path_list=temp_path, topology=self.engine_props['topology'])
```
Comment (ryanmccann1024) [inline src/routing.py:213]: It's not clear the difference in path_len1 and path_len?

File: src/sdn_controller.py [hunk 18]
```diff
@@ -55,10 +56,10 @@ def allocate(self):
         end_slot = self.spectrum_obj.spectrum_props.end_slot
         core_num = self.spectrum_obj.spectrum_props.core_num
         band = self.spectrum_obj.spectrum_props.curr_band
-        self.sdn_props.curr_band = band
 
-        if self.engine_props['guard_slots']:
-            end_slot = end_slot - 1
+        if self.engine_props['guard_slots'] != 0:
```
Comment (ryanmccann1024) [inline src/sdn_controller.py:60]: I think I commented this before but we should definitely have a test for the guard slots if we don't already.

File: src/sdn_controller.py [hunk 19]
```diff
@@ -141,6 +151,38 @@ def _handle_slicing(self, path_list: list, forced_segments: int):
 
             self.sdn_props.is_sliced = False
 
+    def _handle_dynamic_slicing(self, path_list: list, path_index: int, forced_segments: int ):
```
Comment (ryanmccann1024) [inline src/sdn_controller.py:154]: Test for this if we don't have one already.

File: src/sdn_controller.py [hunk 20]
```diff
@@ -141,6 +151,38 @@ def _handle_slicing(self, path_list: list, forced_segments: int):
...
+                else:
+                    self.sdn_props.was_routed = False
+                    self.sdn_props.block_reason = 'congestion'
+                    self.sdn_props.num_trans = 1
+                    if remaining_bw != int(self.sdn_props.bandwidth):
+                        self.release()
+                    self.sdn_props.is_sliced = False
+                    break
+            else:
+                # mod_format = get_path_mod(mods_dict=mods_dict, path_len=path_len)
```
Comment (ryanmccann1024) [inline src/sdn_controller.py:180]: Comment to be removed or not commented.

File: src/snr_measurements.py [hunk 21]
```diff
@@ -305,8 +306,163 @@ def check_xt(self):
...
+    
+    # TODO: update the method based on external resources
+    def check_snr_ext(self, path_index):
+        """
+        Checks the SNR on a single request using the external resources.
+
+        :return: Whether the SNR threshold can be met and SNR value.
+        :rtype: tuple
+        """
+        mod_format_mapping = {
```
Comment (ryanmccann1024) [inline src/snr_measurements.py:318]: These can be included in an "snr_args" file for organizational purposes.

File: src/snr_measurements.py [hunk 22]
```diff
@@ -305,8 +306,163 @@ def check_xt(self):
             resp = cross_talk < self.engine_props['requested_xt'][self.spectrum_props.modulation]
 
         return resp, cross_talk
+    
+    # TODO: update the method based on external resources
+    def check_snr_ext(self, path_index):
```
Comment (ryanmccann1024) [inline src/snr_measurements.py:311]: Test for this method.

File: src/snr_measurements.py [hunk 23]
```diff
@@ -305,8 +306,163 @@ def check_xt(self):
...
+        "16-QAM": 400,
+        "8-QAM": 300,
+        "QPSK": 200,
+        "BPSK": 100
+        }
+        if self.engine_props['multi_fiber']:
+            loaded_data = np.load('MF-USB6014-MF.npy', allow_pickle=True)
+            loaded_data_gsnr = np.load('GSNR-USB6014-MF.npy', allow_pickle=True)
+        else:
+            if self.engine_props['cores_per_link'] == 4:
```
Comment (ryanmccann1024) [inline src/snr_measurements.py:338]: A function in "snr_helpers" would help the organization of this method a lot. Or, have another argument in "snr_args" where it's a tuple mapping to file names. For example: (core_num, cores_per_link): [file1, file2]. The function could then do something like (You'd need to account for a None type, for example the first if statement only cares about cores_per_link):

`file_names_list = snr_args[cores_per_link, core_num]
file_one, file_two = np.load()...
`

File: src/snr_measurements.py [hunk 24]
```diff
@@ -305,8 +306,163 @@ def check_xt(self):
...
+            NotImplementedError(f"Unexpected band: {self.spectrum_props.curr_band}")
+        mod_format = loaded_data[self.route_props.connection_index[0]][slot_index][path_index]
+        SNR_val = loaded_data_gsnr[self.route_props.connection_index[0]][slot_index][path_index]
+        if mod_format == 0:
+            resp = False
+        elif mod_format_mapping[mod_format] == self.spectrum_props.modulation and BW_mapping[self.spectrum_props.modulation] >= int(self.sdn_props.bandwidth):
+            resp = True
+        else:
+            resp = False
+        return resp, SNR_val
```
Comment (ryanmccann1024) [inline src/snr_measurements.py:384]: We return here but have a raise statement afterwards? Will it ever be reached?

File: src/snr_measurements.py [hunk 25]
```diff
@@ -305,8 +306,163 @@ def check_xt(self):
...
+    def check_snr_ext_slicing(self, path_index):
+        """
+        Checks the SNR on a single request using the external resources.
 
-    def handle_snr(self):
+        :return: Whether the SNR threshold can be met and SNR value.
+        :rtype: tuple
+        """
+        mod_format_mapping = {
+        6: "64-QAM",
```
Comment (ryanmccann1024) [inline src/snr_measurements.py:396]: Same comments as above, changing this will help the organization when debugging along with the ability to scale.

File: src/snr_measurements.py [hunk 26]
```diff
@@ -305,8 +306,163 @@ def check_xt(self):
...
+            slot_index = self.spectrum_props.start_slot
+        elif self.spectrum_props.curr_band == 'c':
+            slot_index = self.engine_props['l_band'] + self.spectrum_props.start_slot
+        elif self.spectrum_props.curr_band == 's':
+            slot_index = (self.engine_props['l_band'] + 
+                          self.engine_props['c_band'] + 
+                          self.spectrum_props.start_slot)
+        else:
+            NotImplementedError(f"Unexpected band: {self.spectrum_props.curr_band}")
+        if loaded_data[self.route_props.connection_index[0]][slot_index][path_index] == 0:
```
Comment (ryanmccann1024) [inline src/snr_measurements.py:454]: These functions are very similar, a helper function could be used but it depends on how much they differ.

File: src/spectrum_assignment.py [hunk 27]
```diff
@@ -42,6 +43,7 @@ def _allocate_best_fit(self, channels_list: list):
                     self.spectrum_props.is_free = True
                     self.spectrum_props.start_slot = start_index
                     self.spectrum_props.end_slot = end_index + self.engine_props['guard_slots']
+                    self.spectrum_props.end_slot = end_index
```
Comment (ryanmccann1024) [inline src/spectrum_assignment.py:46]: We should confirm this with the test (I think we do have this test).

File: src/spectrum_assignment.py [hunk 28]
```diff
@@ -107,25 +109,91 @@ def handle_first_last(self, flag: str):
...
-                                         itertools.groupby(enumerate(open_slots_arr), lambda i_x: i_x[0] - i_x[1])]
-                else:
-                    raise NotImplementedError(f'Invalid flag, got: {flag} and expected last_fit or first_fit.')
-
-                self.spec_help_obj.core_num = core_num
-                self.spec_help_obj.curr_band = band_list[band_index]
-                was_allocated = self.spec_help_obj.check_super_channels(open_slots_matrix=open_slots_matrix, flag=flag)
-                if was_allocated:
-                    return
+                for core_arr, core_num in zip(core_matrix, core_list):
```
Comment (ryanmccann1024) [inline src/spectrum_assignment.py:115]: It's hard to follow what's going on here, I think this should be broken up into 3 functions, one main function and two helper functions to help its organization.

File: src/spectrum_assignment.py [hunk 29]
```diff
@@ -107,25 +109,91 @@ def handle_first_last(self, flag: str):
...
+                    open_slots_arr = np.where(core_arr[band_index] == 0)[0]
+
```
Comment (ryanmccann1024) [inline src/spectrum_assignment.py:117]: This will need testing as well.

File: src/spectrum_assignment.py [hunk 30]
```diff
@@ -198,3 +273,36 @@ def get_spectrum(self, mod_format_list: list, slice_bandwidth: str = None):
...
+                    self.spectrum_props.is_free = False
+                    self.sdn_props.block_reason = "xt_threshold"
+                else:
+
+                    self.spectrum_props.modulation = mod_format
+                    self.spectrum_props.xt_cost = snr_val
+                    self.spectrum_props.is_free = True
+                    self.sdn_props.block_reason = None
+                return mod_format, bw
+            else:
```
Comment (ryanmccann1024) [inline src/spectrum_assignment.py:302]: We might want comments here to understand why zeros are being returned or raising a warning/error.

File: tests/test_spectrum_assignment.py [hunk 31]
```diff
@@ -118,6 +123,30 @@ def test_get_spectrum(self):
             self.assertEqual(self.spec_assign.spectrum_props.modulation, 'QPSK',
                              "Expected modulation to be set to 'QPSK'.")
 
+    def test_first_fit_guard_band_zero(self):
```
Comment (ryanmccann1024) [inline tests/test_spectrum_assignment.py:126]: These tests should map to aligned functions in the PR.

File: arg_scripts/routing_args.py [hunk 32]
```diff
@@ -9,6 +9,8 @@ def __init__(self):
         self.paths_matrix = []  # Matrix of potential paths for a single request
         self.mod_formats_matrix = []  # Modulation formats corresponding to each path in paths_matrix
         self.weights_list = []  # Keeping track of one weight of the path (Length, XT, Etc.)
+        self.connection_index = [] # Keeping track of source destination index in precalculated routing
```
Comment (ryanmccann1024) [inline arg_scripts/routing_args.py:12]: Naming guidelines, these are list types.

File: src/routing.py [hunk 33]
```diff
@@ -117,10 +117,14 @@ def find_k_shortest(self):
                 break
             path_len = find_path_len(path_list=path_list, topology=self.engine_props['topology'])
             chosen_bw = self.sdn_props.bandwidth
-            mod_format = get_path_mod(mods_dict=self.engine_props['mod_per_bw'][chosen_bw], path_len=path_len)
-
+            if self.engine_props['pre_calc_mod_selection']:
+                mod_formats = [get_path_mod(mods_dict=self.engine_props['mod_per_bw'][chosen_bw], path_len=path_len)]
```
Comment (ryanmccann1024) [inline src/routing.py:121]: Naming conventions, is this a list or dict?

File: src/routing.py [hunk 34]
```diff
@@ -117,10 +117,14 @@ def find_k_shortest(self):
...
+            else:
+                mod_formats_dict = sort_nested_dict_vals(original_dict=self.sdn_props.mod_formats_dict,
+                                                    nested_key='max_length')
+                mod_formats = list(mod_formats_dict.keys())
```
Comment (ryanmccann1024) [inline src/routing.py:125]: Naming guidelines, type list.

File: src/routing.py [hunk 35]
```diff
@@ -179,6 +183,50 @@ def _init_route_info(self):
...
+                    else:
+                        temp_path = list(path[0][::-1])
+
+                    
+                    temp_path = list(map(str, temp_path))
+                    path_len1 = find_path_len(path_list=temp_path, topology=self.engine_props['topology'])
+                    path_len = item[3][0][cnt]
+                    mod_formats_dict = sort_nested_dict_vals(original_dict=self.sdn_props.mod_formats_dict,
+                                                    nested_key='max_length')
+                    mod_formats = list(mod_formats_dict.keys())
```
Comment (ryanmccann1024) [inline src/routing.py:217]: Naming guidelines.

File: src/sdn_controller.py [hunk 36]
```diff
@@ -141,6 +151,38 @@ def _handle_slicing(self, path_list: list, forced_segments: int):
...
+    def _handle_dynamic_slicing(self, path_list: list, path_index: int, forced_segments: int ):
+        remaining_bw = int(self.sdn_props.bandwidth)
+        path_len = find_path_len(path_list=path_list, topology=self.engine_props['topology'])
+        bw_mod_dict = sort_dict_keys(dictionary=self.engine_props['mod_per_bw'])
+        self.spectrum_obj.spectrum_props.path_list = path_list
+        self.sdn_props.num_trans = 0
+        while remaining_bw > 0:
+            if self.engine_props['fixed_grid']:
+                self.sdn_props.was_routed = True
+                mod_format, bw = self.spectrum_obj.get_spectrum_dynamic_slicing(mod_format_list = [], path_index = path_index)
```
Comment (ryanmccann1024) [inline src/sdn_controller.py:163]: I'm not sure how `bw` passed pylint?

File: src/snr_measurements.py (hunk 22 above)
Comment (ryanmccann1024) [inline src/snr_measurements.py:311]: Path index needs a type.

File: src/snr_measurements.py (hunk 21 above)
Comment (ryanmccann1024) [inline src/snr_measurements.py:318]: Naming, type dict.

File: src/snr_measurements.py [hunk 37]
```diff
@@ -305,8 +306,163 @@ def check_xt(self):
...
+        """
+        mod_format_mapping = {
+        6: "64-QAM",
+        5: "32-QAM",
+        4: "16-QAM",
+        3: "8-QAM",
+        2: "QPSK",
+        1: "BPSK"
+        }
+        BW_mapping = {
```
Comment (ryanmccann1024) [inline src/snr_measurements.py:326]: Naming, type dict.

File: src/snr_measurements.py [hunk 38]
```diff
@@ -305,8 +306,163 @@ def check_xt(self):
...
+        elif self.spectrum_props.curr_band == 'c':
+            slot_index = self.engine_props['l_band'] + self.spectrum_props.start_slot
+        elif self.spectrum_props.curr_band == 's':
+            slot_index = (self.engine_props['l_band'] + 
+                          self.engine_props['c_band'] + 
+                          self.spectrum_props.start_slot)
+        else:
+            NotImplementedError(f"Unexpected band: {self.spectrum_props.curr_band}")
+        mod_format = loaded_data[self.route_props.connection_index[0]][slot_index][path_index]
+        SNR_val = loaded_data_gsnr[self.route_props.connection_index[0]][slot_index][path_index]
```
Comment (ryanmccann1024) [inline src/snr_measurements.py:377]: Snake case naming guidelines.

File: src/snr_measurements.py [hunk 39]
```diff
@@ -318,7 +474,24 @@ def handle_snr(self):
...
+    def handle_snr_dynamic_slicing(self, path_index):
+        """
+        Controls the methods of this class.
+
+        :return: Whether snr is acceptable for allocation or not for a given request and its cost
+        :rtype: tuple
+        """
+        self.num_slots = self.spectrum_props.end_slot - self.spectrum_props.start_slot + 1
+        if self.engine_props['snr_type'] == "snr_e2e_external_resources":
+            mod_format, bw, SNR_val = self.check_snr_ext_slicing(path_index)
```
Comment (ryanmccann1024) [inline src/snr_measurements.py:493]: Snake-case naming.

File: src/spectrum_assignment.py [hunk 40]
```diff
@@ -107,25 +109,91 @@ def handle_first_last(self, flag: str):
...
+
+                    self.spec_help_obj.core_num = core_num
+                    self.spec_help_obj.curr_band = band_list[band_index]
+                    was_allocated = self.spec_help_obj.check_super_channels(open_slots_matrix=open_slots_matrix, flag=flag)
+                    # support FF for the spectrum meet the minimum gsnr requirments
+                    if self.engine_props['cores_per_link'] in [13, 19] and self.engine_props['snr_type'] == 'snr_e2e_external_resources':
+                        snr_result = self.snr_obj.handle_snr_dynamic_slicing(self.sdn_props.path_index)[0]
+                        break_outer_loop = False 
+
+                        if snr_result is None:
```
Comment (ryanmccann1024) [inline src/spectrum_assignment.py:178]: This looks similar to the above code. We can probably add a helper function to interact with both parts with flags as needed.

**Review by ryanmccann1024 — CHANGES_REQUESTED** [review]
I think after we address the comments we should:

- [ ] Confirm SPF, KSP, and LSS work with the same results (using Kimberly's script which I can update)
- [ ] Add needed tests for new functions
- [ ] Map tests to new functions in the PR description
- [ ] Double check team guidelines
- [ ] This can be done first or last, but merge the changes from main onto this branch as there are conflicts

File: arg_scripts/data_args.py [hunk 41]
```diff
@@ -11,19 +11,70 @@
...
 
-    "200": {"QPSK": {"max_length": 2595, "slots_needed": 5}, "16-QAM": {"max_length": 1162, "slots_needed": 3},
-            "64-QAM": {"max_length": 438, "slots_needed": 2}},
+#     "400": {"QPSK": {"max_length": 1298, "slots_needed": 10}, "16-QAM": {"max_length": 581, "slots_needed": 5},
+#             "64-QAM": {"max_length": 219, "slots_needed": 4}}
+# }
 
-    "400": {"QPSK": {"max_length": 1298, "slots_needed": 10}, "16-QAM": {"max_length": 581, "slots_needed": 5},
-            "64-QAM": {"max_length": 219, "slots_needed": 4}}
+ARASH_MOD_ASSUMPTIONS = {
```
[the remark this answers is not in the document]
Comment (arashr88) [reply arg_scripts/data_args.py:31]: Correct, I will add them as examples in the main!

File: arg_scripts/sdn_args.py [hunk 42]
```diff
@@ -7,6 +7,7 @@ class SDNProps:
 
     def __init__(self):
         self.path_list = None  # List of nodes for the current request
+        self.path_index = None # set index of path when set of path are computed
```
[the remark this answers is not in the document]
Comment (arashr88) [reply arg_scripts/sdn_args.py:10]: Precalculated SNRs are structured based on path index

[the remark this answers is not in the document]
Comment (arashr88) [reply arg_scripts/sdn_args.py:10]: removed from args

File: arg_scripts/snr_args.py [hunk 43]
```diff
@@ -22,6 +22,8 @@ def __init__(self):
         self.num_span = None  # Number of span
 
         self.link_dict = None  # Dictionary of links for calculating various metrics
+        self.mod_format_mapping_dict = None # Dictionary of Modulation formats for precalculated SNR
+        self.bw_mapping_dict = None # Dictionary of Modulation formats to calculate the supported bit ratre for fixed grid
```
[the remark this answers is not in the document]
Comment (arashr88) [reply arg_scripts/snr_args.py:26]: fixed

File: arg_scripts/stats_args.py [hunk 44]
```diff
@@ -13,12 +13,17 @@ def __init__(self):
         self.block_bw_dict = dict()  # Block per bandwidth
         self.block_reasons_dict = {'distance': None, 'congestion': None, 'xt_threshold': None}  # Block reasons
         self.sim_block_list = list()  # List of blocking probabilities per simulation
+        self.sim_bit_rate_block_list = list()  # List of blocking probabilities per simulation
```
[the remark this answers is not in the document]
Comment (arashr88) [reply arg_scripts/stats_args.py:16]: fixed

File: arg_scripts/snr_args.py (hunk 43 above)
[the remark this answers is not in the document]
Comment (arashr88) [reply arg_scripts/snr_args.py:26]: Conversion between MF and bit rate, and vice versa in Fixed-grid

File: bash_scripts/run_sim.sh (hunk 5 above)
Comment (arashr88) [reply bash_scripts/run_sim.sh:31]: Yeah, I updated based on new OS

File: src/routing.py (hunk 14 above)
Comment (arashr88) [reply src/routing.py:204]: updated based on self.engine_props['k_paths']

File: src/routing.py (hunk 15 above)
Comment (arashr88) [reply src/routing.py:199]: node

File: src/routing.py (hunk 16 above)
Comment (arashr88) [reply src/routing.py:200]: input file includes multiple information so I considered the item.

File: src/routing.py (hunk 17 above)
Comment (arashr88) [reply src/routing.py:213]: removed

File: src/snr_measurements.py (hunk 24 above)
Comment (arashr88) [reply src/snr_measurements.py:384]: removed

File: src/spectrum_assignment.py (hunk 30 above)
Comment (arashr88) [reply src/spectrum_assignment.py:302]: 0 means lightpath can not be accommodated

File: arg_scripts/routing_args.py [hunk 45]
```diff
@@ -9,13 +9,15 @@ def __init__(self):
         self.paths_matrix = []  # Matrix of potential paths for a single request
         self.mod_formats_matrix = []  # Modulation formats corresponding to each path in paths_matrix
         self.weights_list = []  # Keeping track of one weight of the path (Length, XT, Etc.)
+        self.path_index = [] # Keeping track of path index in precalculated routing
```
Comment (ryanmccann1024) [inline arg_scripts/routing_args.py:12]: Variable naming.

File: arg_scripts/snr_args.py [hunk 46]
```diff
@@ -24,6 +24,17 @@ def __init__(self):
         self.link_dict = None  # Dictionary of links for calculating various metrics
         self.mod_format_mapping_dict = {6: "64-QAM", 5: "32-QAM", 4: "16-QAM", 3: "8-QAM", 2: "QPSK", 1: "BPSK"} # Dictionary of Modulation formats for precalculated SNR
         self.bw_mapping_dict = {"64-QAM": 600, "32-QAM": 500, "16-QAM": 400, "8-QAM": 300, "QPSK": 200, "BPSK": 100} # Dictionary of Modulation formats to calculate the supported bit rate for fixed grid
+        self.file_mapping_dict = {
+            (4, 7): {'mf': 'MF-USB6014-MCF4-C2.npy', 'gsnr': 'GSNR-USB6014-MCF4-C2.npy'},
```
Comment (arashr88) [inline arg_scripts/snr_args.py:28]: key should be changed to (2,4)

**Review by ryanmccann1024 — APPROVED** [review]
